src/control/server/ctl_storage_rpc.go

```diff
@@ -43,87 +42,88 @@ const (
 	msgNvmeFormatSkip = "NVMe format skipped on instance %d as SCM format did not complete"
 )
 
-// newState creates, populates and returns ResponseState in addition
-// to logging any err.
-func newState(log logging.Logger, status ctlpb.ResponseStatus, errMsg string, infoMsg string,
-	contextMsg string) *ctlpb.ResponseState {
+// newResponseState creates, populates and returns ResponseState.
+func newResponseState(inErr error, badStatus ctlpb.ResponseStatus, infoMsg string) *ctlpb.ResponseState {
+	rs := new(ctlpb.ResponseState)
+	rs.Info = infoMsg
 
-	state := &ctlpb.ResponseState{
-		Status: status, Error: errMsg, Info: infoMsg,
-	}
-
-	if errMsg != "" {
-		// TODO: is this necessary, maybe not?
-		log.Error(contextMsg + ": " + errMsg)
+	if inErr != nil {
+		rs.Status = badStatus
+		rs.Error = inErr.Error()
 	}
 
-	return state
+	return rs
 }
 
-func (c *StorageControlService) doNvmePrepare(req *ctlpb.PrepareNvmeReq) (resp *ctlpb.PrepareNvmeResp) {
-	resp = &ctlpb.PrepareNvmeResp{}
-	msg := "Storage Prepare NVMe"
+// doNvmePrepare issues prepare request and returns response.
+func (c *StorageControlService) doNvmePrepare(req *ctlpb.PrepareNvmeReq) *ctlpb.PrepareNvmeResp {
 	_, err := c.NvmePrepare(bdev.PrepareRequest{
 		HugePageCount: int(req.GetNrhugepages()),
 		TargetUser:    req.GetTargetuser(),
 		PCIWhitelist:  req.GetPciwhitelist(),
 		ResetOnly:     req.GetReset_(),
 	})
 
-	if err != nil {
-		resp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_NVME, err.Error(), "", msg)
-		return
-	}
+	pnr := new(ctlpb.PrepareNvmeResp)
+	pnr.State = newResponseState(err, ctlpb.ResponseStatus_CTL_ERR_NVME, "")
 
-	resp.State = newState(c.log, ctlpb.ResponseStatus_CTL_SUCCESS, "", "", msg)
-	return
+	return pnr
 }
 
-func (c *StorageControlService) doScmPrepare(pbReq *ctlpb.PrepareScmReq) (pbResp *ctlpb.PrepareScmResp) {
-	pbResp = &ctlpb.PrepareScmResp{}
-	msg := "Storage Prepare SCM"
+// newPrepareScmResp sets protobuf SCM prepare response with results.
+func newPrepareScmResp(inResp *scm.PrepareResponse, inErr error) (*ctlpb.PrepareScmResp, error) {
+	var info string
+	var err error
+	outResp := new(ctlpb.PrepareScmResp)
 
-	scmState, err := c.GetScmState()
-	if err != nil {
-		pbResp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(), "", msg)
-		return
-	}
-	c.log.Debugf("SCM state before prep: %s", scmState)
+	defer func() {
+		if err != nil {
+			return
+		}
+		outResp.State = newResponseState(inErr, ctlpb.ResponseStatus_CTL_ERR_SCM, info)
+	}()
```

Comment:
The defer function makes things a little more harder to reason about here. Since there are only two points where we want to set the state in this function, I'd be fine with just repeating that part.

Reply:
have re-factored to remove derring the creation of state